tail: tests with piped input are failing randomly with broken pipe on macos when no output is expected

Here's the reference to the original discussion:

> I did some more testing on macOS.
> The failures will also sometimes (10-20% of the time) happen on the `main` branch.
> However, they will only happen inside the test framework, i.e. this will never trigger a failure:
> ```bash
> for i in {1..100}; do echo "a\nb\n" | target/debug/tail -n +3; done
> ```
> but this will fail regularly:
> ```rust
> let test_string = "a\nb\n";
> new_ucmd!()
>     .args(&["-n", "+3"])
>     .pipe_in(test_string)
>     .succeeds()
>     .no_stdout()
>     .no_stderr();
> ```
> 

_Originally posted by @jhscheer in https://github.com/uutils/coreutils/pull/3874#discussion_r961947237_

EDIT:
Here's the exact message from one of the failing tests during the CICD #8262 (https://github.com/uutils/coreutils/actions/runs/2939929799) run:
```
2022-08-27T20:38:23.8017460Z failures:
2022-08-27T20:38:23.8017780Z 
2022-08-27T20:38:23.8019590Z ---- test_tail::pipe_tests::test_pipe_when_negative_lines_option_given_no_newline_at_eof stdout ----
2022-08-27T20:38:23.8020010Z current_directory_resolved: 
2022-08-27T20:38:23.8021360Z run: /Users/runner/work/coreutils/coreutils/target/debug/coreutils tail -n 0
2022-08-27T20:38:23.8022960Z thread 'test_tail::pipe_tests::test_pipe_when_negative_lines_option_given_no_newline_at_eof' panicked at 'failed to write to stdin of child: Broken pi
pe (os error 32)', tests/common/util.rs:1115:21
2022-08-27T20:38:23.8023550Z 
2022-08-27T20:38:23.8023700Z 
2022-08-27T20:38:23.8023940Z failures:
2022-08-27T20:38:23.8024440Z     test_tail::pipe_tests::test_pipe_when_negative_lines_option_given_no_newline_at_eof
2022-08-27T20:38:23.8024840Z 
2022-08-27T20:38:23.8025270Z test result: FAILED. 1904 passed; 1 failed; 31 ignored; 0 measured; 0 filtered out; finished in 109.75s
2022-08-27T20:38:23.8025710Z 
2022-08-27T20:38:23.8097270Z error: test failed, to rerun pass '--test tests'
2022-08-27T20:38:23.8208890Z ##[error]The process '/Users/runner/.cargo/bin/cargo' failed with exit code 101
```

Looking at the code around `tests/common/util.rs:1115` the child `target/debug/coreutils tail -n 0` seams to have closed its stdin already when trying to write to its stdin. But maybe I'm wrong. Strange that it happens only on macos.

--- a/tests/common/util.rs
+++ b/tests/common/util.rs
@@ -1828,4 +1828,22 @@ mod tests {
             println!("TEST SKIPPED (cannot run inside CI)");
         }
     }
+
+    // This error was first detected when running tail so tail is used here but
+    // should fail with any command that takes piped input.
+    // See also https://github.com/uutils/coreutils/issues/3895
+    #[test]
+    fn test_when_piped_input_then_no_broken_pipe() {
+        let ts = TestScenario::new("tail");
+        for i in 0..10000 {
+            dbg!(i);
+            let test_string = "a\nb\n";
+            ts.ucmd()
+                .args(&["-n", "0"])
+                .pipe_in(test_string)
+                .succeeds()
+                .no_stdout()
+                .no_stderr();
+        }
+    }
 }

--- a/tests/common/util.rs
+++ b/tests/common/util.rs
@@ -17,7 +17,7 @@ use std::env;
 use std::ffi::CString;
 use std::ffi::OsStr;
 use std::fs::{self, hard_link, File, OpenOptions};
-use std::io::{Read, Result, Write};
+use std::io::{BufWriter, Read, Result, Write};
 #[cfg(unix)]
 use std::os::unix::fs::{symlink as symlink_dir, symlink as symlink_file};
 #[cfg(windows)]
@@ -1105,13 +1105,14 @@ impl UCommand {
         }
 
         if let Some(ref input) = self.bytes_into_stdin {
-            let write_result = child
+            let child_stdin = child
                 .stdin
                 .take()
-                .unwrap_or_else(|| panic!("Could not take child process stdin"))
-                .write_all(input);
+                .unwrap_or_else(|| panic!("Could not take child process stdin"));
+            let mut writer = BufWriter::new(child_stdin);
+            let result = writer.write_all(input);
             if !self.ignore_stdin_write_error {
-                if let Err(e) = write_result {
+                if let Err(e) = result {
                     panic!("failed to write to stdin of child: {}", e);
                 }
             }

